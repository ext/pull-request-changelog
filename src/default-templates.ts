const defaultHeaderTemplate = "## {{ title }}";
const defaultFooterTemplate = "";
const defaultMessageTemplate = `{{> header }}


{{#if changelog}}
Based on commits in this Pull Request this will create a **{{ bump }}** release and the following entries will be added to the changelog:",

---
{{ changelog }}
---

If this is not correct you can amend the commit message(s).

{{else}}
No commits contributing to a release was found, no changelog entries will be added for this Pull Request.


If this PR should be included in a release amend the commit message(s) to use:

- \`fix:\` for a patch version or
- \`feat:\` for minor version.

{{/if}}
{{>footer}}
`;

/**
 * @internal
 */
export const defaultTemplates = {
	message: defaultMessageTemplate,
	header: defaultHeaderTemplate,
	footer: defaultFooterTemplate,
} as const;
