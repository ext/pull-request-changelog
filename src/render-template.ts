import handlebars from "handlebars";

/**
 * @internal
 */
export interface Templates {
	message: string;
	header: string;
	footer: string;
}

/**
 * @internal
 */
export interface TemplateData {
	/** changelog entries or `false` if no commits yielding changelog entries was found */
	changelog: string | false;
	/** configured title */
	title: string;
	/** what version bump this would generate for the next release */
	bump: "major" | "minor" | "patch";
}

/**
 * Render handlebar templates.
 *
 * @internal
 */
export function renderTemplate(template: Templates, data: TemplateData): string {
	const fn = handlebars.compile(template.message);
	handlebars.registerPartial("header", template.header);
	handlebars.registerPartial("footer", template.footer);
	return fn(data);
}
