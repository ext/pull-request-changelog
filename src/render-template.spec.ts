import { defaultTemplates } from "./default-templates";
import { renderTemplate } from "./render-template";

it("should render template with changelog", async () => {
	expect.assertions(1);
	const changelog = [
		"### Features",
		"",
		" * foo",
		" * bar",
		"",
		"### Bug fixes",
		"",
		" * baz",
		"",
	].join("\n");
	const output = await renderTemplate(defaultTemplates, {
		changelog,
		title: "CHANGELOG",
		bump: "minor",
	});
	expect(output).toMatchSnapshot();
});

it("should render template with no changelog", async () => {
	expect.assertions(1);
	const changelog = false;
	const output = await renderTemplate(defaultTemplates, {
		changelog,
		title: "CHANGELOG",
		bump: "minor",
	});
	expect(output).toMatchSnapshot();
});
