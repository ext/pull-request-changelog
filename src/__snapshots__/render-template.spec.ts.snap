// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render template with changelog 1`] = `
"## CHANGELOG

Based on commits in this Pull Request this will create a **minor** release and the following entries will be added to the changelog:

---
### Features

 * foo
 * bar

### Bug fixes

 * baz

---

If this is not correct you can amend the commit message(s).

"
`;

exports[`should render template with no changelog 1`] = `
"## CHANGELOG

No commits contributing to a release was found, no changelog entries will be added for this Pull Request.

If this PR should be included in a release amend the commit message(s) to use:

- \`fix:\` for a patch version or
- \`feat:\` for minor version.

"
`;
